### PR TITLE
test script for deepseek

### PR DESCRIPTION
**Pull request by shuningjin**
# Description

**Main change**
onboard test for deepseek2-16b, plan to run full workflow in XLML on v5p-8
- `end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh`: ckpt conversion, logit check, train, finetune, decode

onboard test for deepseek3-671b, plan to run step 2 in XLML on v5p-128
- `end_to_end/tpu/deepseek/v3-671b/1_test_deepseek.sh`
- `end_to_end/tpu/deepseek/v3-671b/2_test_deepseek.sh`: logit check (scan), decode (unscan). Note we leave out training and finetuning, as is covered by ubench nightly test.

XLML PR: https://github.com/GoogleCloudPlatform/ml-auto-solutions/pull/1112

FIXES: b/423057893

**Other change**
- Minor format change to other test scripts. 
- `max_logging.log` behavior is changed due to [PR#2873](https://github.com/AI-Hypercomputer/maxtext/pull/2873), need to set absl.logging.level to make logs visible for multiple scripts

# Tests

preparation
- The checkpoints uploaded
  - gs://maxtext-deepseek/deepseek2-16b (hf-bf16)
  - gs://maxtext-deepseek/deepseek3-671b (hf-bf16, scanned, unscanned)
- golden logits uploaded
  - gs://maxtext-test-assets/golden_data_deepseek2-16b.jsonl
  - gs://maxtext-test-assets/golden_data_deepseek3-671b.jsonl

`bash end_to_end/tpu/deepseek/v3-671b/2_test_deepseek.sh`
- xpk on v5p-128: https://paste.googleplex.com/5746766743994368
- log: https://cloudlogging.app.goo.gl/38jsiXx6y3miv7TL7

`bash end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh`
- xpk on v5p-8: https://paste.googleplex.com/6732499990478848
- log, 30min: https://cloudlogging.app.goo.gl/J7SA4eLjMvjNyNGm8

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: end_to_end/tpu/deepseek/Run_DeepSeek.md [hunk 1]
```diff
@@ -151,7 +151,7 @@ python3 -m MaxText.decode src/MaxText/configs/base.yml \
     max_target_length=1024 \
     tokenizer_type=huggingface \
     tokenizer_path=deepseek-ai/DeepSeek-V3 \
-    attention=flash \
```
Comment (RissyRan) [inline end_to_end/tpu/deepseek/Run_DeepSeek.md:154]: Thanks!

File: end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh [hunk 2]
```diff
@@ -3,49 +3,79 @@
...
+
+# Run fine-tuning - megablox implementation
+python3 -m MaxText.train "${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/MaxText}/"configs/base.yml base_output_directory=${BASE_OUTPUT_PATH} run_name=megablox_fine_tuning model_name=${MODEL_NAME} tokenizer_type=huggingface tokenizer_path=${TOKENIZER_PATH} dataset_path=${DATASET_PATH} enable_checkpointing=true async_checkpointing=false load_parameters_path=${SCANNED_CKPT_PATH} scan_layers=True attention=flash sparse_matmul=True megablox=True dtype=bfloat16 weight_dtype=bfloat16 per_device_batch_size=4 steps=5 max_target_length=1024 ici_fsdp_parallelism=1 ici_expert_parallelism=4 checkpoint_storage_concurrent_gb=1024
+
+# Run supervised fine-tuning - megablox implementation
```
Comment (RissyRan) [inline end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh]: Let's remove it if it's comment?

File: end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh [hunk 3]
```diff
@@ -3,49 +3,79 @@
...
+
+# Run pre-training - megablox implementation
+python3 -m MaxText.train "${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/MaxText}/"configs/base.yml base_output_directory=${BASE_OUTPUT_PATH} run_name=megablox_pre_training model_name=${MODEL_NAME} tokenizer_type=huggingface tokenizer_path=${TOKENIZER_PATH} dataset_type=synthetic enable_checkpointing=false attention=flash sparse_matmul=True megablox=True dtype=bfloat16 weight_dtype=bfloat16 per_device_batch_size=4 steps=5 max_target_length=1024 ici_fsdp_parallelism=4
+
+# Run fine-tuning - megablox implementation
```
Comment (RissyRan) [inline end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh]: Let's update megablox to be tokamax gmm. We may deprecate megablox sometime later, and most of workloads are using tokamax gmm now.

File: end_to_end/tpu/deepseek/v3-671b/2_test_deepseek.sh [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+GOLDEN_LOGITS_DISK_LOCATION="/deps/src/MaxText/test_assets/golden_data_${MODEL_NAME}.jsonl"
+if [ ! -f "${GOLDEN_LOGITS_DISK_LOCATION}" ]; then
+  GOLDEN_LOGITS_PATH="gs://maxtext-test-assets/golden_data_${MODEL_NAME}.jsonl"
+  GOLDEN_LOGITS_DISK_LOCATION=/tmp/golden_data.jsonl
+  gcloud storage cp ${GOLDEN_LOGITS_PATH} ${GOLDEN_LOGITS_DISK_LOCATION}
+fi
+
+python3 -m tests.forward_pass_logit_checker ${MAXTEXT_PKG_DIR}/configs/base.yml base_output_directory=${BASE_OUTPUT_PATH} run_name=forward_logits_check load_parameters_path=${SCANNED_CKPT_PATH} scan_layers=true attention=dot_product per_device_batch_size=1 model_name=${MODEL_NAME} max_prefill_predict_length=4 max_target_length=4 async_checkpointing=false sparse_matmul=false ici_fsdp_parallelism=1 ici_expert_parallelism=-1 checkpoint_storage_concurrent_gb=1024 weight_dtype=float32 dtype=float32 activations_in_float32=true matmul_precision=highest float32_logits=true float32_qk_product=true --golden_logits_path=${GOLDEN_LOGITS_DISK_LOCATION} --atol=1.5 --rtol=1.5 --max_kl_div=0.1
+
+# Run pre-training - megablox implementation
```
Comment (RissyRan) [inline end_to_end/tpu/deepseek/v3-671b/2_test_deepseek.sh]: similar comments above.

File: end_to_end/tpu/llama4/2_test_llama4.sh [hunk 5]
```diff
@@ -14,25 +14,26 @@
 set -ex
 idx=$(date +%Y-%m-%d)
 
+# By default, we'll use "llama4-17b-16e"
+if [ -z "${MODEL_VARIATION}" ]; then
```
Comment (RissyRan) [inline end_to_end/tpu/llama4/2_test_llama4.sh:18]: Do we need to set this from XL ML? or we cold direct `export MODEL_VARIATION="llama4-17b-16e"` without if/else

File: end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh [hunk 6]
```diff
@@ -39,7 +41,7 @@ python3 -m MaxText.decode "${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/Max
...
+python3 -m tests.forward_pass_logit_checker "${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/MaxText}/"configs/base.yml base_output_directory=${BASE_OUTPUT_PATH} load_parameters_path=${UNSCANNED_CKPT_PATH} run_name=matmul_forward_pass_test per_device_batch_size=1 model_name=mixtral-8x7b tokenizer_path="${MAXTEXT_ASSETS_ROOT:-${MAXTEXT_PKG_DIR:-${MAXTEXT_REPO_ROOT:-$PWD}/src/MaxText/assets}}"/tokenizer.mistral-v1 ici_tensor_parallelism=1 ici_fsdp_parallelism=-1 max_prefill_predict_length=11 max_target_length=11 dataset_type=synthetic dtype=float32 megablox=False sparse_matmul=False scan_layers=false --token_size=4 --atol=3 --rtol=1 --max_kl_div=2e-3
```
Comment (RissyRan) [inline end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh]: Do you know if `--atol=3 --rtol=1` is still needed?

File: end_to_end/tpu/deepseek/v3-671b/2_test_deepseek.sh (hunk 4 above)
Comment (shuningjin) [reply end_to_end/tpu/deepseek/v3-671b/2_test_deepseek.sh]: I assume you mean remove (rather than update with tokamax)? Removed train/finetune/sft

File: end_to_end/tpu/llama4/2_test_llama4.sh (hunk 5 above)
Comment (shuningjin) [reply end_to_end/tpu/llama4/2_test_llama4.sh:18]: For other models, we have separate test scripts for each variant. For llama4, this serves as the united script. Do we need to account for Maverick as well?

File: end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh (hunk 2 above)
Comment (shuningjin) [reply end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh]: Removed this unused command for sft

File: end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh (hunk 6 above)
Comment (shuningjin) [reply end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh]: Removed `--atol=3 --rtol=1`, instead using criteria `--max_kl_div=2e-3`.

File: end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh (hunk 3 above)
Comment (shuningjin) [reply end_to_end/tpu/deepseek/v2-16b/test_deepseek.sh]: Thanks for clarifying the MoE strategies! I updated to tokamax_gmm for training and decoding for deepseek

File: end_to_end/tpu/llama4/2_test_llama4.sh (hunk 5 above)
Comment (RissyRan) [reply end_to_end/tpu/llama4/2_test_llama4.sh:18]: I see. I think we should do something similar to GPT-OSS models. By passing the type instead of duplicate the scripts. No need to be in this PR.

**Review by RissyRan — APPROVED** [review]
Thanks for the updates!

**Comment by shuningjin** [comment]
For deepseek script, updated moe strategy to tokamax_gmm for training and decoding. Tests pass:
- ds3: https://cloudlogging.app.goo.gl/onW4CPgVkZ5SZhBVA
- ds2: https://cloudlogging.app.goo.gl/z8u5cMrpykniX7ji6

File: end_to_end/tpu/deepseek/Run_DeepSeek.md (hunk 1 above)
Comment (shuningjin) [reply end_to_end/tpu/deepseek/Run_DeepSeek.md:154]: This command line is for decode.py. decode on tpu does not support flash attention, but only dot_product
https://github.com/AI-Hypercomputer/maxtext/blob/24cdc86538d5cfb9f836b60266e0a02152af6dbb/src/MaxText/layers/attention_op.py#L869-L880

File: end_to_end/tpu/deepseek/v3-671b/2_test_deepseek.sh [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+fi
+BASE_OUTPUT_PATH=${BASE_OUTPUT_PATH%/}
+echo using BASE_OUTPUT_PATH = ${BASE_OUTPUT_PATH}
+
+# Step 2: 
+# We define the checkpoint paths. This way it is easier to use these paths in the `train.py` and `decode.py` commands
+# export SCANNED_CKPT_PATH=${BASE_OUTPUT_PATH}/scanned/0/items
+# export UNSCANNED_CKPT_PATH=${BASE_OUTPUT_PATH}/unscanned/0/items
+SCANNED_CKPT_PATH=gs://maxtext-deepseek/deepseek3-671b/2025-03-31/scanned/0/items
+UNSCANNED_CKPT_PATH=gs://maxtext-deepseek/deepseek3-671b/2025-03-31/unscanned/0/items
```
Comment (parambole) [inline end_to_end/tpu/deepseek/v3-671b/2_test_deepseek.sh:39]: If this script is intended to run daily and pick up artifacts generated by 1_test_deepseek.sh (which uses $(date +%Y-%m-%d-%H-%M)), this hardcoding will break the flow or cause the test to validate stale artifacts rather than the newly converted ones.

If the intention is to always test against a known-good "Golden" checkpoint, then this is fine, but it might be worth adding a comment clarifying that we are intentionally skipping the artifacts generated by Step 1.

Comment (shuningjin) [reply end_to_end/tpu/deepseek/v3-671b/2_test_deepseek.sh:39]: Thank you for highlighting this! 
- For ds3, we intentionally use a golden checkpoint in Step 2. Step 1 takes very long and is skipped for daily test. conversion is covered by ds2.
- I have added a comment for clarifying.

File: end_to_end/tpu/llama4/2_test_llama4.sh (hunk 5 above)
Comment (shuningjin) [reply end_to_end/tpu/llama4/2_test_llama4.sh:18]: Different variant may need customization in tests, for instance
- forward logit check criteria
- parallelism as they are typically run on different devices (e.g., v5p-8 vs. v5p-128)

I would prefer keeping disentangled scripts for different variants when possible.

Actually, for the llama4 case, the logit criteria is set for scout. this might not work for maverick. Need to follow this up.
